bundles/org.openhab.voice.actiontemplatehli/src/main/java/org/openhab/voice/actiontemplatehli/internal/ActionTemplateInterpreter.java

```diff
@@ -0,0 +1,1159 @@
+    }
+
+    @Override
+    public Set<Locale> getSupportedLocales() {
+        // all are supported
+        return Set.of();
+    }
+
+    @Override
+    public Set<String> getSupportedGrammarFormats() {
+        return Set.of();
+    }
+
+    @Override
+    public String interpret(Locale locale, String words) throws InterpretationException {
+        if (words.isEmpty()) {
+            new InterpretationException(config.unhandledMessage);
+        }
+        try {
+            if (config.lowerText) {
+                words = words.toLowerCase(locale);
+            }
+            var info = getNLPInfo(words);
+            if (info.tokens.length == 0) {
+                logger.warn("no tokens produced; aborting");
+                return config.failMessage;
+            }
+            String response = processAction(words, checkActionConfigs(words, info.tokens, info.tags, info.lemmas));
+            if (response == null) {
+                logger.debug("silent mode; no response");
+                return "";
+            }
+            logger.debug("response: {}", response);
+            return response;
+        } catch (IOException e) {
+            logger.debug("IOException while interpreting: {}", e.getMessage(), e);
+            var message = e.getMessage();
+            throw new InterpretationException(message != null ? message : "Unknown error");
+        } catch (RuntimeException e) {
+            var message = e.getMessage();
+            logger.debug("RuntimeException while interpreting: {}", e.getMessage(), e);
+            throw new InterpretationException(message != null ? message : "Unknown error");
+        }
+    }
+
+    private @Nullable String processAction(String words, @Nullable NLPInterpretationResult result)
+            throws InterpretationException, IOException {
+        if (result != null) {
+            if (!result.actionConfig.read) {
+                return sendItemCommand(result.targetItem, words, result.actionConfig, result.placeholderValues);
+            } else {
+                return readItemState(result.targetItem, result.actionConfig);
+            }
+        } else {
+            throw new InterpretationException(config.unhandledMessage);
+        }
+    }
+
+    public NLPInfo getNLPInfo(String text) throws IOException {
+        logger.debug("Processing: '{}'", text);
+        var nlpInfo = cacheMap.get(text);
+        if (nlpInfo == null) {
+            var tokens = tokenizeText(text);
+            var tags = languagePOSTagging(tokens);
+            var lemmas = languageLemmatize(tokens, tags);
+            nlpInfo = new NLPInfo(tokens, lemmas, tags);
+        } else {
+            logger.debug("resolved from cache");
+        }
+        logger.debug("tokens: {}", List.of(nlpInfo.tokens));
+        logger.debug("tags: {}", List.of(nlpInfo.tags));
+        logger.debug("lemmas: {}", List.of(nlpInfo.lemmas));
+        var finalNlpInfo = nlpInfo;
+        cacheMap.put(text, () -> finalNlpInfo);
```

Comment:
Looks strange: You provide a supplier something that will return always the same data ?

Reply:
True, it was due to type requirements, for me it will be enough to put the value but the ExpiringCacheMap seems to require this. Not sure if there is an alternative class.

Reply:
Now, that is not my comment. 
If U am correct, this parameter should be the supplier to get a fresh new value.